### DISCUSSION TITLE
WHO'S idea what this ???

### DISCUSSION THREAD

**Discussion by angryguy-win** [ChrisTitusTech/linutil#778] [Q&A] [answered]
WHO'S idea was this to make all of the scripts Posix  complaint and why ??? 
What did you think the benefit would be by doing this?.
Just out of curiosity which distro, of Linux does not run Bash ?...
I understand not every one every like bash, or uses bash.. fair point 
In recent video I have heard Chris complain about the Quality of scripts coming in PR's ?.. 
I would really like to know before I answer why I think this is a very bad idea and will hurt the project in the long run...
and I know your first response will be that I'm crazy and this upsets you and that you had a good reason fine, explain the benefits ?..
 Because I have been looking at doing or better said converting code to Posix compliant and ummmm not seeing the benefits AT ALL portability?, can you please help me understand  ?..

local variables security
arrays, associative arrays
pattern matching, regex 
conditional expressions
built  in functions
debugging 
I just want to understand that all nothing more nothing less....
I understand using it for the Common scripts i get it but .... every single Linux distro out there that I know of can change to any shell at any time can change in an instant ... You create the tool chain for the install.. clean up when your done...

**Comment by lj3954** [comment]
I can forward this one to POSIX purist @nnyyxxxx.

**Comment by ghost** [comment, marked as answer]
bash scripts are slow and messy and it's very easy to write scripts without bashisms, there's no reason to write scripts in bash and i don't see the benefit of doing that. the current scripts in linutil are okay they can be better but that's some of the contributors faults for not testing. Most of the script additions in Linutil were made by @guruswarupa if you have any inquiries i would rather you talk to him about it.

**Comment by angryguy-win** [comment]
pass the buck...  other then telling me you don't enjoy or like Bash,  the only thing you had to add was slow and messy how?.. 
Elaborate ?.. @nnyyxxxx

**Reply by ghost** [reply]
i thought it was common knowledge that bash specific features slow scripts down, even just running a script without bashisms will reduce overhead by a large margin.

Especially if paired up with an optimized shell like for instance "dash"

if your sh is indeed linked to dash and the script contains no bashisms and it includes the sh shebang then the script will run a large margin faster than it would compared to a bloated interactive shell like bash with multiple bashisms in the script.

**Comment by koibtw** [comment]
Just one thing: I get that you proudly wear the name of Angry Guy Win but stop here for a second. Who are you? What is your problem? Nyx is doing a **lot** of work for this project. And I mean a **lot**. Just look at the contributors page. What did you do for linutil? Because everytime I see your name you're complaining.

**Comment by angryguy-win** [comment]
I'm  not complaining I ask simple question, I want to understand, I' am person just like any body else asking question, I have no problem, Your the one that seams upset, did I upset you?. I'm sure he's does lots for LinUtil never doubted it?..

And then you attack me , by saying all's I ever do is complain as like you know who I am, you don't know me, any more then I know you... I'm being perfectly civil and Asking questions which I have every right too... 
This is a discussion board is not ?..
Where people can ask question and learn, instead you choice to try be little me... as my opinion doesn't matter because you believe that I do nothing but complain... 
I all so don't have to agree @nnyyxxxx  I still respect his opinion, he think it faster he think it portable, he doesn't like bash, personally I like bash and  can write very complex scripts whit it.. I don't expect @nnyyxxxx  to use bash any more then I want the use Sh...

**Comment by angryguy-win** [comment]
The differences between Bash and sh (Bourne Shell) can be significant in various aspects, including speed, performance, usability, security, and POSIX compliance. Here’s a breakdown of these factors:
1. Speed and Performance
Startup Time:
sh: Generally has a faster startup time because it is simpler and has fewer features to load.
Bash: May have a slightly longer startup time due to its additional features and built-in functions.
Execution Speed:
The execution speed of scripts can vary based on the complexity of the script and the features used. For simple scripts, the difference in execution speed is often negligible. However, Bash's advanced features (like arrays and built-in string manipulation) can lead to more efficient scripts in some cases.
2. Usability
User Interface:
sh: Provides a basic command-line interface with limited features for user interaction.
Bash: Offers a more user-friendly experience with features like command-line editing, history, tab completion, and better error messages.
Scripting Features:
sh: Supports basic scripting capabilities, including control structures (if, for, while) and functions.
Bash: Includes advanced features such as arrays, associative arrays, extended pattern matching, and more robust string manipulation, making it easier to write complex scripts.
3. Security
Security Features:
Both shells can be secure if used properly, but Bash has more features that can be leveraged for security, such as better handling of user input and more robust error handling.
sh: Simplicity can be an advantage in security, as fewer features mean fewer potential vulnerabilities.
Bash: The complexity of features can introduce security risks if not managed correctly, especially with features like command substitution and variable expansion.
4. POSIX Compliance
POSIX Compliance:
sh: The original Bourne Shell is the basis for the POSIX standard for shell scripting, making it fully compliant.
Bash: While Bash is largely POSIX compliant, it also includes many extensions and features that are not part of the POSIX standard. This means that while Bash can run POSIX-compliant scripts, scripts that use Bash-specific features may not run in a pure POSIX shell.
5. Why Use One Over the Other?
Use sh When:
You need maximum portability across different Unix-like systems.
You are writing simple scripts that do not require advanced features.
You want to ensure strict POSIX compliance.
Use Bash When:
You need advanced scripting capabilities and features (like arrays, string manipulation, etc.).
You are working in a controlled environment where Bash is guaranteed to be available (e.g., most Linux distributions).
You want a more user-friendly command-line experience with features like command history and tab completion.

Conclusion
The choice between Bash and sh depends on your specific needs and the environment in which your scripts will run. If portability and strict compliance are your priorities, sh is the better choice. If you need advanced features and a more user-friendly experience, Bash is the way to go.
Questions
Q1: Are there specific features or functionalities you are looking for in your scripts that might influence your choice between Bash and sh?
Q2: Do you have a particular environment or system in mind where you plan to deploy your scripts?
Q3: Would you like assistance in refactoring existing scripts to leverage the strengths of either shell?

**Reply by ghost** [reply]
is this LLM generated?

**Reply by ghost** [reply]
not sure you understand that dash + posix will always be much faster than just bash

**Reply by ghost** [reply]
not to mention the portability that you gain with writing in posix

**Reply by ghost** [reply]
and anyways everything that you can do in bash specific syntax you can do with posix, it's usually just longer, BUT its always going to be more portable and more performing than the bash version

**Reply by ghost** [reply]
https://wiki.archlinux.org/title/Dash

read this for more information

**Reply by angryguy-win** [reply]
yes that was LLM generated ..
I did not says, I do not understand that dash + posix is not faster I have no doubt that it is... Not disputing it...

But where talking about shell scripts here is 200ns or ok 2 minutes going to make a difference on a shell script that what a few hundred lines maybe a few thousand lines?..
I'm being serious, here and I'm not saying that your opinion is any less or better then mine on the contrary if you believe the best possible solution is speed over all else great maybe you have more to add, I don't know this is why I'm asking...
I understand completely being posix complaint for the common base script.. 

But every single script in the script folder hence are nt each script stand alone from the project it's self and LinUtil just being the collector of those scripts ?...  was that not the intent for the project to make it easier for general user's to share idea and scripts...

**Reply by ghost** [reply]
> yes that was LLM generated .. I did not says, I do not understand that dash + posix is not faster I have no doubt that it is... Not disputing it...
> 
> But where talking about shell scripts here is 200ns or ok 2 minutes going to make a difference on a shell script that what a few hundred lines maybe a few thousand lines?.. I'm being serious, here and I'm not saying that your opinion is any less or better then mine on the contrary if you believe the best possible solution is speed over all else great maybe you have more to add, I don't know this is why I'm asking... I understand completely being posix complaint for the common base script..
> 
> But every single script in the script folder hence are nt each script stand alone from the project it's self and LinUtil just being the collector of those scripts ?... was that not the intent for the project to make it easier for general user's to share idea and scripts...

why write scripts in bash when posix is faster and more portable? there is no benefit to writing scripts in bash compared to posix

**Reply by angryguy-win** [reply]
I agree whit the faster part, but the portable part bash has what 2 Distro's that have issues whit bash Alpine and busybox, and Yes I get that you don't like bash... Thank you for your time good sir.. I think Sh and dash ect..  all have a place some where, I just think it foolish to throw away tool when all tools have there purpose... agree to disagree... 
I don't discredit Dash, I'm sure it lovely never used it my self maybe. I will give it a try now... and see for my self ..

I write in bash because I feel comfortable whit it.. and If I where more confident whit RUST as I'm still learning maybe some day I can contribute too.. in college we where forced to learn Java and C# and other various web dev stuff and in linux class we where show bash scripting... so that what I write... 
and as other read this... they can make up there own minds...
